src/librustc/mir/interpret/error.rs

```diff
@@ -438,139 +438,165 @@ pub enum UnsupportedOpInfo<'tcx> {
 impl fmt::Debug for UnsupportedOpInfo<'tcx> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         use UnsupportedOpInfo::*;
+        #[rustfmt::skip] // rustfmt and long matches do not go well together
```

Comment:
Imo the diff doesn't seem like that much of an improvement to justify this... I'm concerned about people starting to use this in various places throughout the compiler to fit their personal tastes rather than having a uniform style.

Reply:
I tend to agree with @Centril here that I would like to avoid `#[rustfmt::skip]`. I'd rather find ways to structure the code that make rustfmt happy, or else open issues on rustfmt for extreme cases.

Reply:
(Tweaking rustfmt.toml seems ok too, though I'd prefer to avoid even that when we can.)

Reply:
Uh... for me this diff is the difference between "unreadable garbage" and "reasonable code". Like, without this PR, the `write!` is formatted in one of *three* different ways depending on the length of the message and the phase of the moon. Sometimes it's entirely in the same line as the pattern, sometimes it is in a line on its own, and sometimes the `write!` is on the line of the pattern but its arguments are in separate lines. How is that *not* horrible for anyone? It's inconsistent, it's hard to parse visually. The entire point of rustfmt, I thought, is to get consistent formatting, but clearly it is doing the opposite here (and for `match`es in general).

I opened a rustfmt issue at https://github.com/rust-lang/rustfmt/issues/3995. But until that gets fixed, I don't see any good reason to let a tool's deficiency (rustfmt overall works fairly poorly on `match` arms) dictate the format.

Reply:
Also Cc @Mark-Simulacrum, whose suggestion it was to use `rustfmt::skip` on this file.

Reply:
You may not like this option, but you could also rewrite the code to use this pattern:

let val;
match input {
    pat_1 => {
        val = expr_1;
    }
    pat_2 => {
        val = expr_2;
    }
    ...
}
```

At least, my informal experiments during the compiler team meeting indicated that using a statement would force `rustfmt` to keep the braces, and once you have the braces, it tended to keep the newlines that would make the `write!` inovcations match up.

Reply:
So in order to please rustfmt, we'll change code to be less idiomatic (statements instead of expressions)? That seems... odd? rustfmt isn't some God-given axiom that we have to blindly follow, is it?

Reply:
Fwiw, I think that rewriting beautiful functional-looking code into weird unidiomatic imperative code would be taking things too far. Maybe it's not optimal, but I don't think the style that is being replaced in this diff is *that* egregious that we need to use `val = expr_1;`.

> rustfmt isn't some God-given axiom that we have to blindly follow, is it?

No, it's just useful to be able to format-on-save anywhere, not caring about style and have it fixed for you, as well as a uniform reading experience, and reducing pure formatting changes in `git blame` over time.

Reply:
>  it's just useful to be able to format-on-save anywhere

I appreciate that's useful for folks who went through the hoops of setting that up. Even as someone who didn't, I appreciate the consistent formatting of function signatures, the sorted imports, stuff like that.

But when considering `match`es like this, I honestly don't understand how you can call this a "uniform reading experience" -- there's nothing uniform about how rustfmt formats this match. As far as I am concerned, the formatting of this match after rustfmt is literally worse than anything I ever saw anywhere in the rustc codebase before rustfmt. The way it formats matches *will* make it much harder for me to work on and review this code-base, due to how it degrades readability. I appreciate that is subjective, but the part where it is very inconsistent and using different styles to do the exact same thing is objectively neither uniform nor consistent.

I think the style in this diff is by far egregious enough to justify `rustfmt::skip`. I also don't get the principled opposition to that: we don't usually make ourselves slaves to our own tools, do we? Tidy has had exceptions that were and are applied where reasonable. Not allowing reasonable exceptions would be quite unprecedented. In fact, I count around 30 occurrences of `rustfmt::skip` in the rustc repo already (excluding tools and tests). Many of them are for some sort of table or another (but in some sense, this `match` *is* a table), but [this one is not](https://github.com/rust-lang/rust/blob/f48e576756cd0c360e5522974fc8d5867b439092/src/liballoc/boxed.rs#L595). And that makes sense -- rustfmt is a great tool, but like all tools it has its limitations. We should apply the tool where it makes sense, but not mindlessly submit ourselves to whatever the tool happens to do.

Reply:
I'll just note that sometimes I find `let val; ...; val = expr` *more* readable.

(Not when its mutable state; that's almost never more readable. I'm talking about lazily initialized state here.)

Why do I find it more readable? Well, sometimes the `let val = { ... }; ends up with something where the result expression is so far from the binding that it doesn't fit in my editor's view at once, which means I've effectively lost the relationship (or at least have to keep in in my mental cache).

Reply:
> Why do I find it more readable? Well, sometimes the `let val = { ... }; ends up with something where the result expression is so far from the binding that it doesn't fit in my editor's view at once, which means I've effectively lost the relationship (or at least have to keep in in my mental cache).

Split your functions Felix, so that "doesn't fit" doesn't happen :P

Reply:
> Split your functions Felix, so that "doesn't fit" doesn't happen :P

Sure, when that's an option, it can work well. And I suppose you might also suggest that I ensure that the enums I process never have more than four or five variants.

but sometimes you have to play the hand you are dealt, and I'm glad that I have `let x; ... x = ...;` as a tool in my back pocket, even you all aren't willing to use it yourself.

Reply:
> And I suppose you might also suggest that I ensure that the enums I process never have more than four or five variants.

Not really, but you can split `match` arms into separate functions when the `match` gets too long, e.g. as in https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_typeck/check/expr.rs.html#213-289 or https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_typeck/check/pat.rs.html#119-219 which were unreadable messes before I split those functions apart.